client/rest-high-level/src/main/java/org/elasticsearch/client/security/support/expressiondsl/parser/RoleMapperExpressionParser.java

```diff
@@ -0,0 +1,220 @@
+            break;
+        case "groups":
+            builder = FieldExpressionBuilder.builder(GroupsFieldExpression.class);
+            break;
+        case "dn":
+            builder = FieldExpressionBuilder.builder(DnFieldExpression.class);
+            break;
+        case "metadata":
+            builder = FieldExpressionBuilder.builder(MetadataFieldExpression.class);
+            builder.withKey(parsedFieldName);
+            break;
+        default:
+            throw new ElasticsearchParseException("failed to parse field expression, unexpected field name [{}]", fieldName);
+        }
+        return builder;
+    }
+
+    private String fieldName(String objectName, XContentParser parser) throws IOException {
+        if (parser.nextToken() != XContentParser.Token.FIELD_NAME) {
+            throw new ElasticsearchParseException("failed to parse rules expression. object [{}] does not contain any fields", objectName);
+        }
+        String parsedFieldName = parser.currentName();
+        return parsedFieldName;
+    }
+
+    private List<RoleMapperExpression> parseExpressionArray(ParseField field, XContentParser parser, boolean allowExcept)
+            throws IOException {
+        parser.nextToken(); // parseArray requires that the parser is positioned
+                            // at the START_ARRAY token
+        return parseArray(field, parser, p -> parseRulesObject(field.getPreferredName(), p, allowExcept));
+    }
+
+    private <T> List<T> parseArray(ParseField field, XContentParser parser, CheckedFunction<XContentParser, T, IOException> elementParser)
+            throws IOException {
+        final XContentParser.Token token = parser.currentToken();
+        if (token == XContentParser.Token.START_ARRAY) {
+            List<T> list = new ArrayList<>();
+            while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                list.add(elementParser.apply(parser));
+            }
+            return list;
+        } else {
+            throw new ElasticsearchParseException("failed to parse rules expression. field [{}] requires an array", field);
+        }
+    }
+
+    private Object parseFieldValue(XContentParser parser) throws IOException {
+        switch (parser.currentToken()) {
+        case VALUE_STRING:
+            return parser.text();
+
+        case VALUE_BOOLEAN:
+            return parser.booleanValue();
+
+        case VALUE_NUMBER:
+            return parser.longValue();
+
+        case VALUE_NULL:
+            return null;
+
+        default:
+            throw new ElasticsearchParseException("failed to parse rules expression. expected a field value but found [{}] instead", parser
+                    .currentToken());
+        }
+    }
+
+    public interface Fields {
+        ParseField ANY = new ParseField("any");
+        ParseField ALL = new ParseField("all");
+        ParseField EXCEPT = new ParseField("except");
+        ParseField FIELD = new ParseField("field");
```

Comment:
We shouldn't have both these fields and also the `NAME` fields on the various expression types.